File: frontend/src/services/template_service.ts
/**
 * Service for Document Templates API
 */

import api from './api';
import type {
  DocumentTemplate,
  TemplateListItem,
  TemplateCreateData,
  TemplateUpdateData,
  TemplateVariable,
  VariableCreateData,
  VariableUpdateData,
  TemplateType,
  GenerateDocumentRequest,
  GeneratedDocumentResponse,
  TemplateAutocompleteResponse,
  VariableValue,
} from '@/types/template';

interface PaginatedResponse<T> {
  items: T[];
  total: number;
  skip: number;
  limit: number;
}

const BASE_URL = '/templates';
const GENERATION_URL = '/document-generation';

/**
 * Template CRUD operations
 */
export const templateService = {
  /**
   * Get list of templates with pagination and filtering
   */
  async getTemplates(params?: {
    skip?: number;
    limit?: number;
    template_type?: TemplateType;
    is_active?: boolean;
  }): Promise<PaginatedResponse<TemplateListItem>> {
    const searchParams = new URLSearchParams();

    if (params?.skip !== undefined) searchParams.append('skip', params.skip.toString());
    if (params?.limit !== undefined) searchParams.append('limit', params.limit.toString());
    if (params?.template_type) searchParams.append('template_type', params.template_type);
    if (params?.is_active !== undefined) searchParams.append('is_active', params.is_active.toString());

    const query = searchParams.toString();
    const url = query ? `${BASE_URL}?${query}` : BASE_URL;

    const response = await api.get<PaginatedResponse<TemplateListItem>>(url);
    return response.data;
  },

  /**
   * Get template by ID with variables
   */
  async getTemplate(id: number): Promise<DocumentTemplate> {
    const response = await api.get<DocumentTemplate>(`${BASE_URL}/${id}`);
    return response.data;
  },

  /**
   * Create new template
   */
  async createTemplate(data: TemplateCreateData, file: File): Promise<DocumentTemplate> {
    const formData = new FormData();
    formData.append('file', file);

    const searchParams = new URLSearchParams();
    searchParams.append('name', data.name);
    searchParams.append('code', data.code);
    searchParams.append('template_type', data.template_type);
    if (data.description) searchParams.append('description', data.description);
    if (data.default_output_format) searchParams.append('default_output_format', data.default_output_format);

    const response = await api.post<DocumentTemplate>(`${BASE_URL}?${searchParams.toString()}`, formData, {
      headers: { 'Content-Type': 'multipart/form-data' },
    });
    return response.data;
  },

  /**
   * Update template metadata
   */
  async updateTemplate(id: number, data: TemplateUpdateData): Promise<DocumentTemplate> {
    const response = await api.put<DocumentTemplate>(`${BASE_URL}/${id}`, data);
    return response.data;
  },

  /**
   * Update template file
   */
  async updateTemplateFile(id: number, file: File): Promise<DocumentTemplate> {
    const formData = new FormData();
    formData.append('file', file);
    const response = await api.put<DocumentTemplate>(`${BASE_URL}/${id}/file`, formData, {
      headers: { 'Content-Type': 'multipart/form-data' },
    });
    return response.data;
  },

  /**
   * Delete template
   */
  async deleteTemplate(id: number): Promise<void> {
    await api.delete(`${BASE_URL}/${id}`);
  },

  // =========================================================================
  // Variable operations
  // =========================================================================

  /**
   * Get template variables
   */
  async getVariables(templateId: number): Promise<TemplateVariable[]> {
    const response = await api.get<TemplateVariable[]>(`${BASE_URL}/${templateId}/variables`);
    return response.data;
  },

  /**
   * Create variable
   */
  async createVariable(templateId: number, data: VariableCreateData): Promise<TemplateVariable> {
    const response = await api.post<TemplateVariable>(`${BASE_URL}/${templateId}/variables`, data);
    return response.data;
  },

  /**
   * Update variable
   */
  async updateVariable(
    templateId: number,
    variableId: number,
    data: VariableUpdateData
  ): Promise<TemplateVariable> {
    const response = await api.put<TemplateVariable>(
      `${BASE_URL}/${templateId}/variables/${variableId}`,
      data
    );
    return response.data;
  },

  /**
   * Delete variable
   */
  async deleteVariable(templateId: number, variableId: number): Promise<void> {
    await api.delete(`${BASE_URL}/${templateId}/variables/${variableId}`);
  },

  /**
   * Reorder variables
   */
  async reorderVariables(templateId: number, variableIds: number[]): Promise<TemplateVariable[]> {
    const response = await api.put<TemplateVariable[]>(
      `${BASE_URL}/${templateId}/variables/reorder`,
      variableIds
    );
    return response.data;
  },
};

/**
 * Document generation operations
 */
export const documentGenerationService = {
  /**
   * Get autocomplete data for template
   */
  async getAutocomplete(
    templateId: number,
    performanceId?: number
  ): Promise<TemplateAutocompleteResponse> {
    const params = new URLSearchParams();
    if (performanceId) params.append('performance_id', performanceId.toString());

    const query = params.toString();
    const url = query
      ? `${GENERATION_URL}/templates/${templateId}/autocomplete?${query}`
      : `${GENERATION_URL}/templates/${templateId}/autocomplete`;

    const response = await api.get<TemplateAutocompleteResponse>(url);
    return response.data;
  },

  /**
   * Generate preview (returns blob)
   */
  async generatePreview(
    templateId: number,
    variables: VariableValue[],
    performanceId?: number
  ): Promise<Blob> {
    const response = await api.post(`${GENERATION_URL}/preview`, {
      template_id: templateId,
      performance_id: performanceId,
      variables,
    }, {
      responseType: 'blob',
    });
    return response.data;
  },

  /**
   * Generate document
   */
  async generateDocument(request: GenerateDocumentRequest): Promise<GeneratedDocumentResponse> {
    const response = await api.post<GeneratedDocumentResponse>(`${GENERATION_URL}/generate`, request);
    return response.data;
  },

  /**
   * Quick generate for performance
   */
  async quickGenerateForPerformance(
    templateId: number,
    performanceId: number,
    outputFormat: 'docx' | 'pdf' = 'docx'
  ): Promise<GeneratedDocumentResponse> {
    const response = await api.post<GeneratedDocumentResponse>(
      `${GENERATION_URL}/templates/${templateId}/generate-for-performance/${performanceId}?output_format=${outputFormat}`,
      {}
    );
    return response.data;
  },
};

File: backend/app/services/template_service.py
"""
Сервис для работы с шаблонами документов.

Бизнес-логика для:
- CRUD операций с шаблонами
- Управления переменными шаблонов
- Валидации
"""
from typing import Any

from fastapi import UploadFile
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.exceptions import AlreadyExistsError, NotFoundError, ValidationError
from app.models.document_template import (
    DocumentTemplate,
    DocumentTemplateVariable,
    TemplateType,
)
from app.repositories.template_repository import (
    TemplateRepository,
    TemplateVariableRepository,
)
from app.schemas.document_template import (
    TemplateCreate,
    TemplateUpdate,
    TemplateVariableCreate,
    TemplateVariableUpdate,
)
from app.services.minio_service import MinioService


# Разрешённые расширения для шаблонов
ALLOWED_TEMPLATE_EXTENSIONS = {".docx"}


class TemplateService:
    """Сервис для работы с шаблонами документов."""

    def __init__(self, session: AsyncSession) -> None:
        """
        Инициализация сервиса.

        Args:
            session: Асинхронная сессия БД
        """
        self._session = session
        self._template_repo = TemplateRepository(session)
        self._variable_repo = TemplateVariableRepository(session)
        self._minio = MinioService()

    # =========================================================================
    # Template CRUD
    # =========================================================================

    async def get_template(self, template_id: int) -> DocumentTemplate:
        """
        Получить шаблон по ID.

        Args:
            template_id: ID шаблона

        Returns:
            Экземпляр шаблона

        Raises:
            NotFoundError: Если шаблон не найден
        """
        template = await self._template_repo.get_by_id(template_id)
        if template is None:
            raise NotFoundError(f"Шаблон с ID {template_id} не найден")
        return template

    async def get_template_by_code(self, code: str) -> DocumentTemplate:
        """
        Получить шаблон по коду.

        Args:
            code: Код шаблона

        Returns:
            Экземпляр шаблона

        Raises:
            NotFoundError: Если шаблон не найден
        """
        template = await self._template_repo.get_by_code(code)
        if template is None:
            raise NotFoundError(f"Шаблон с кодом '{code}' не найден")
        return template

    async def get_templates(
        self,
        *,
        skip: int = 0,
        limit: int = 100,
        theater_id: int | None = None,
        template_type: TemplateType | None = None,
        is_active: bool | None = None,
    ) -> tuple[list[DocumentTemplate], int]:
        """
        Получить список шаблонов с фильтрацией.

        Args:
            skip: Пропустить записей
            limit: Максимум записей
            theater_id: Фильтр по театру
            template_type: Фильтр по типу
            is_active: Фильтр по активности

        Returns:
            Кортеж (список шаблонов, общее количество)
        """
        templates = await self._template_repo.get_all(
            skip=skip,
            limit=limit,
            theater_id=theater_id,
            template_type=template_type,
            is_active=is_active,
        )

        total = await self._template_repo.count(
            theater_id=theater_id,
            template_type=template_type,
            is_active=is_active,
        )

        return templates, total

    async def create_template(
        self,
        data: TemplateCreate,
        file: UploadFile,
        theater_id: int | None = None,
        user_id: int | None = None,
    ) -> DocumentTemplate:
        """
        Создать новый шаблон.

        Args:
            data: Данные для создания
            file: Файл шаблона (DOCX)
            theater_id: ID театра
            user_id: ID пользователя-создателя

        Returns:
            Созданный шаблон

        Raises:
            AlreadyExistsError: Если код уже существует
            ValidationError: Если файл невалидный
        """
        # Проверка уникальности кода
        if await self._template_repo.code_exists(data.code):
            raise AlreadyExistsError(f"Шаблон с кодом '{data.code}' уже существует")

        # Валидация файла
        await self._validate_template_file(file)

        # Загрузка файла в MinIO
        file_path = await self._upload_template_file(file, data.code)

        # Создание шаблона
        template_data = {
            "name": data.name,
            "code": data.code,
            "description": data.description,
            "file_path": file_path,
            "file_name": file.filename,
            "template_type": data.template_type,
            "default_output_format": data.default_output_format,
            "settings": data.settings,
            "theater_id": theater_id,
            "created_by_id": user_id,
            "updated_by_id": user_id,
        }

        template = await self._template_repo.create(template_data)

        # Создание переменных, если переданы
        if data.variables:
            variables_data = [v.model_dump() for v in data.variables]
            await self._variable_repo.bulk_create(template.id, variables_data)

            # Перезагрузка шаблона с переменными
            template = await self._template_repo.get_by_id(template.id)

        await self._session.commit()
        return template

    async def update_template(
        self,
        template_id: int,
        data: TemplateUpdate,
        user_id: int | None = None,
    ) -> DocumentTemplate:
        """
        Обновить шаблон.

        Args:
            template_id: ID шаблона
            data: Данные для обновления
            user_id: ID пользователя

        Returns:
            Обновлённый шаблон

        Raises:
            NotFoundError: Если шаблон не найден
        """
        template = await self.get_template(template_id)

        update_data = data.model_dump(exclude_unset=True)
        if user_id:
            update_data["updated_by_id"] = user_id

        template = await self._template_repo.update(template, update_data)
        await self._session.commit()
        return template

    async def update_template_file(
        self,
        template_id: int,
        file: UploadFile,
        user_id: int | None = None,
    ) -> DocumentTemplate:
        """
        Обновить файл шаблона.

        Args:
            template_id: ID шаблона
            file: Новый файл шаблона
            user_id: ID пользователя

        Returns:
            Обновлённый шаблон
        """
        template = await self.get_template(template_id)

        # Валидация файла
        await self._validate_template_file(file)

        # Удаление старого файла
        try:
            await self._minio.delete_file(template.file_path)
        except Exception:
            pass  # Игнорируем ошибки удаления

        # Загрузка нового файла
        file_path = await self._upload_template_file(file, template.code)

        update_data = {
            "file_path": file_path,
            "file_name": file.filename,
        }
        if user_id:
            update_data["updated_by_id"] = user_id

        template = await self._template_repo.update(template, update_data)
        await self._session.commit()
        return template

    async def delete_template(self, template_id: int) -> None:
        """
        Удалить шаблон.

        Args:
            template_id: ID шаблона

        Raises:
            NotFoundError: Если шаблон не найден
            ValidationError: Если шаблон системный
        """
        template = await self.get_template(template_id)

        if template.is_system:
            raise ValidationError("Нельзя удалить системный шаблон")

        # Удаление файла из MinIO
        try:
            await self._minio.delete_file(template.file_path)
        except Exception:
            pass

        await self._template_repo.delete(template)
        await self._session.commit()

    # =========================================================================
    # Variable CRUD
    # =========================================================================

    async def get_variable(self, variable_id: int) -> DocumentTemplateVariable:
        """Получить переменную по ID."""
        variable = await self._variable_repo.get_by_id(variable_id)
        if variable is None:
            raise NotFoundError(f"Переменная с ID {variable_id} не найдена")
        return variable

    async def get_template_variables(
        self,
        template_id: int,
    ) -> list[DocumentTemplateVariable]:
        """Получить все переменные шаблона."""
        # Проверка существования шаблона
        await self.get_template(template_id)
        return await self._variable_repo.get_by_template_id(template_id)

    async def create_variable(
        self,
        template_id: int,
        data: TemplateVariableCreate,
    ) -> DocumentTemplateVariable:
        """
        Создать переменную шаблона.

        Args:
            template_id: ID шаблона
            data: Данные переменной

        Returns:
            Созданная переменная

        Raises:
            NotFoundError: Если шаблон не найден
            AlreadyExistsError: Если переменная с таким именем уже есть
        """
        # Проверка существования шаблона
        await self.get_template(template_id)

        # Проверка уникальности имени в шаблоне
        existing = await self._variable_repo.get_by_name(template_id, data.name)
        if existing:
            raise AlreadyExistsError(
                f"Переменная '{data.name}' уже существует в шаблоне"
            )

        variable_data = data.model_dump()
        variable_data["template_id"] = template_id

        variable = await self._variable_repo.create(variable_data)
        await self._session.commit()
        return variable

    async def update_variable(
        self,
        variable_id: int,
        data: TemplateVariableUpdate,
    ) -> DocumentTemplateVariable:
        """Обновить переменную."""
        variable = await self.get_variable(variable_id)

        # Проверка уникальности имени, если меняется
        if data.name and data.name != variable.name:
            existing = await self._variable_repo.get_by_name(
                variable.template_id, data.name
            )
            if existing:
                raise AlreadyExistsError(
                    f"Переменная '{data.name}' уже существует в шаблоне"
                )

        update_data = data.model_dump(exclude_unset=True)
        variable = await self._variable_repo.update(variable, update_data)
        await self._session.commit()
        return variable

    async def delete_variable(self, variable_id: int) -> None:
        """Удалить переменную."""
        variable = await self.get_variable(variable_id)
        await self._variable_repo.delete(variable)
        await self._session.commit()

    async def reorder_variables(
        self,
        template_id: int,
        variable_ids: list[int],
    ) -> list[DocumentTemplateVariable]:
        """
        Изменить порядок переменных.

        Args:
            template_id: ID шаблона
            variable_ids: Список ID переменных в новом порядке

        Returns:
            Список переменных в новом порядке
        """
        # Проверка существования шаблона
        await self.get_template(template_id)

        await self._variable_repo.reorder(template_id, variable_ids)
        await self._session.commit()

        return await self._variable_repo.get_by_template_id(template_id)

    # =========================================================================
    # Helper Methods
    # =========================================================================

    async def _validate_template_file(self, file: UploadFile) -> None:
        """Валидация файла шаблона."""
        if not file.filename:
            raise ValidationError("Имя файла не указано")

        # Проверка расширения
        ext = "." + file.filename.rsplit(".", 1)[-1].lower()
        if ext not in ALLOWED_TEMPLATE_EXTENSIONS:
            raise ValidationError(
                f"Недопустимый формат файла. Разрешены: {', '.join(ALLOWED_TEMPLATE_EXTENSIONS)}"
            )

        # Проверка размера (10 MB макс для шаблонов)
        max_size = 10 * 1024 * 1024
        content = await file.read()
        await file.seek(0)

        if len(content) > max_size:
            raise ValidationError("Размер файла превышает 10 MB")

    async def _upload_template_file(
        self,
        file: UploadFile,
        template_code: str,
    ) -> str:
        """
        Загрузить файл шаблона в MinIO.

        Args:
            file: Файл для загрузки
            template_code: Код шаблона

        Returns:
            Путь к файлу в хранилище
        """
        content = await file.read()
        await file.seek(0)

        # Формируем путь: templates/{code}/{filename}
        file_path = f"templates/{template_code.lower()}/{file.filename}"

        await self._minio.upload_file(
            file_path=file_path,
            content=content,
            content_type="application/vnd.openxmlformats-officedocument.wordprocessingml.document",
        )

        return file_path
